File: sankofa_app/lib/services/api_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import '../config/app_config.dart';
import 'api_exception.dart';
import 'auth_service.dart';

class ApiClient {
  ApiClient({http.Client? httpClient, AuthService? authService})
      : _httpClient = httpClient ?? http.Client(),
        _authService = authService ?? AuthService();

  final http.Client _httpClient;
  final AuthService _authService;

  Future<dynamic> get(String path, {Map<String, dynamic>? queryParameters}) {
    final uri = AppConfig.resolve(path, queryParameters);
    return _request('GET', uri);
  }

  Future<dynamic> post(String path, {Object? body}) {
    final uri = AppConfig.resolve(path);
    return _request('POST', uri, body: body);
  }

  Future<dynamic> postMultipart(
    String path, {
    Map<String, String>? fields,
    List<http.MultipartFile>? files,
  }) {
    final uri = AppConfig.resolve(path);
    return _multipartRequest('POST', uri, fields: fields, files: files);
  }

  static const Duration _defaultTimeout = Duration(seconds: 20);

  Future<dynamic> _request(
    String method,
    Uri uri, {
    Object? body,
    bool retrying = false,
  }) async {
    final headers = <String, String>{'Accept': 'application/json'};
    final accessToken = await _authService.getAccessToken(allowRefresh: false);
    if (accessToken != null && accessToken.isNotEmpty) {
      headers['Authorization'] = 'Bearer $accessToken';
    }
    if (body != null) {
      headers['Content-Type'] = 'application/json';
    }

    http.Response response;
    final payload = body == null
        ? null
        : body is String
            ? body
            : jsonEncode(body);

    try {
      switch (method) {
        case 'POST':
          response = await _httpClient
              .post(uri, headers: headers, body: payload)
              .timeout(_defaultTimeout);
          break;
        case 'GET':
        default:
          response = await _httpClient
              .get(uri, headers: headers)
              .timeout(_defaultTimeout);
      }
    } on TimeoutException {
      throw ApiException(
        'The request timed out. Please check your connection and try again.',
      );
    } on SocketException catch (error) {
      throw ApiException(
        'Unable to reach the server. Please check your internet connection.',
        details: {'reason': error.message},
      );
    } on http.ClientException catch (error) {
      throw ApiException('Unable to reach the server. Please try again.', details: {'reason': error.message});
    }

    if (response.statusCode == 401) {
      final shouldRetry = !retrying && await _authService.refreshAccessToken();
      if (shouldRetry) {
        return _request(method, uri, body: body, retrying: true);
      }
      throw ApiException('Your session has expired. Please sign in again.', statusCode: 401);
    }

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw _buildException(response);
    }

    if (response.body.isEmpty) {
      return null;
    }

    try {
      return jsonDecode(utf8.decode(response.bodyBytes));
    } on FormatException {
      return response.body;
    }
  }

  Future<dynamic> _multipartRequest(
    String method,
    Uri uri, {
    Map<String, String>? fields,
    List<http.MultipartFile>? files,
    bool retrying = false,
  }) async {
    final request = http.MultipartRequest(method, uri);
    request.headers['Accept'] = 'application/json';

    final accessToken = await _authService.getAccessToken(allowRefresh: false);
    if (accessToken != null && accessToken.isNotEmpty) {
      request.headers['Authorization'] = 'Bearer $accessToken';
    }

    if (fields != null && fields.isNotEmpty) {
      request.fields.addAll(fields);
    }
    if (files != null && files.isNotEmpty) {
      request.files.addAll(files);
    }

    http.StreamedResponse streamedResponse;
    try {
      streamedResponse = await request.send().timeout(_defaultTimeout);
    } on TimeoutException {
      throw ApiException(
        'The request timed out. Please check your connection and try again.',
      );
    } on SocketException catch (error) {
      throw ApiException(
        'Unable to reach the server. Please check your internet connection.',
        details: {'reason': error.message},
      );
    } on http.ClientException catch (error) {
      throw ApiException(
        'Unable to reach the server. Please try again.',
        details: {'reason': error.message},
      );
    }

    final response = await http.Response.fromStream(streamedResponse);

    if (response.statusCode == 401) {
      final shouldRetry = !retrying && await _authService.refreshAccessToken();
      if (shouldRetry) {
        return _multipartRequest(method, uri, fields: fields, files: files, retrying: true);
      }
      throw ApiException('Your session has expired. Please sign in again.', statusCode: 401);
    }

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw _buildException(response);
    }

    if (response.bodyBytes.isEmpty) {
      return null;
    }

    try {
      return jsonDecode(utf8.decode(response.bodyBytes));
    } on FormatException {
      return response.body;
    }
  }

  ApiException _buildException(http.Response response) {
    Map<String, dynamic>? details;
    String message = 'Request failed with status ${response.statusCode}.';

    if (response.bodyBytes.isNotEmpty) {
      try {
        final decoded = jsonDecode(utf8.decode(response.bodyBytes));
        if (decoded is Map<String, dynamic>) {
          details = decoded;
          final detail = decoded['detail'] ?? decoded['message'];
          if (detail is String && detail.isNotEmpty) {
            message = detail;
          }
        }
      } catch (_) {
        // ignore parse errors
      }
    }

    return ApiException(message, statusCode: response.statusCode, details: details);
  }
}

File: sankofa_frontend/src/lib/apiClient.ts
/**
 * API Client
 * Mirrors the mobile app's ApiClient with automatic token refresh
 */

import AppConfig from './config';
import { ApiException } from './apiException';
import { authService } from '../services/authService';

const DEFAULT_TIMEOUT = 20000; // 20 seconds

class ApiClient {
  private async request<T>(
    method: 'GET' | 'POST',
    path: string,
    options?: {
      body?: unknown;
      queryParams?: Record<string, string | number | boolean>;
      retrying?: boolean;
    }
  ): Promise<T> {
    const url = AppConfig.resolve(path, options?.queryParams);
    const headers: Record<string, string> = {
      'Accept': 'application/json',
    };

    // Get access token without refresh on first attempt
    const accessToken = await authService.getAccessToken(false);
    if (accessToken) {
      headers['Authorization'] = `Bearer ${accessToken}`;
    }

    if (options?.body) {
      headers['Content-Type'] = 'application/json';
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), DEFAULT_TIMEOUT);

    try {
      const response = await fetch(url, {
        method,
        headers,
        body: options?.body ? JSON.stringify(options.body) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      // Handle 401 with token refresh
      if (response.status === 401) {
        if (!options?.retrying) {
          const refreshed = await authService.refreshAccessToken();
          if (refreshed) {
            return this.request<T>(method, path, { ...options, retrying: true });
          }
        }
        throw new ApiException('Your session has expired. Please sign in again.', 401);
      }

      // Handle non-2xx responses
      if (response.status < 200 || response.status >= 300) {
        throw await this.buildException(response);
      }

      // Handle empty response
      const text = await response.text();
      if (!text) {
        return null as T;
      }

      // Try to parse JSON
      try {
        return JSON.parse(text) as T;
      } catch {
        return text as T;
      }
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof ApiException) {
        throw error;
      }

      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new ApiException('The request timed out. Please check your connection and try again.');
        }
        throw new ApiException('Unable to reach the server. Please check your internet connection.');
      }

      throw new ApiException('An unexpected error occurred.');
    }
  }

  private async buildException(response: Response): Promise<ApiException> {
    let message = `Request failed with status ${response.status}.`;
    let details: Record<string, unknown> | undefined;

    try {
      const text = await response.text();
      if (text) {
        const decoded = JSON.parse(text);
        if (typeof decoded === 'object' && decoded !== null) {
          details = decoded as Record<string, unknown>;
          const detail = decoded.detail || decoded.message;
          if (typeof detail === 'string' && detail) {
            message = detail;
          } else if (Array.isArray(detail) && detail.length > 0 && typeof detail[0] === 'string') {
            message = detail[0];
          } else {
            // Try to extract first string value from response
            for (const value of Object.values(decoded)) {
              if (typeof value === 'string' && value) {
                message = value;
                break;
              }
              if (Array.isArray(value) && value.length > 0 && typeof value[0] === 'string') {
                message = value[0];
                break;
              }
            }
          }
        }
      }
    } catch {
      // Ignore parse errors
    }

    return new ApiException(message, response.status, details);
  }

  async get<T>(path: string, queryParams?: Record<string, string | number | boolean>): Promise<T> {
    return this.request<T>('GET', path, { queryParams });
  }

  async post<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>('POST', path, { body });
  }

  async postMultipart<T>(
    path: string,
    fields?: Record<string, string>,
    files?: Array<{ name: string; file: File }>
  ): Promise<T> {
    const url = AppConfig.resolve(path);
    const headers: Record<string, string> = {
      'Accept': 'application/json',
    };

    const accessToken = await authService.getAccessToken(false);
    if (accessToken) {
      headers['Authorization'] = `Bearer ${accessToken}`;
    }

    const formData = new FormData();
    if (fields) {
      Object.entries(fields).forEach(([key, value]) => {
        formData.append(key, value);
      });
    }
    if (files) {
      files.forEach(({ name, file }) => {
        formData.append(name, file);
      });
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), DEFAULT_TIMEOUT);

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers,
        body: formData,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (response.status === 401) {
        const refreshed = await authService.refreshAccessToken();
        if (refreshed) {
          // Retry once after refresh
          return this.postMultipart<T>(path, fields, files);
        }
        throw new ApiException('Your session has expired. Please sign in again.', 401);
      }

      if (response.status < 200 || response.status >= 300) {
        throw await this.buildException(response);
      }

      const text = await response.text();
      if (!text) {
        return null as T;
      }

      try {
        return JSON.parse(text) as T;
      } catch {
        return text as T;
      }
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof ApiException) {
        throw error;
      }

      if (error instanceof Error && error.name === 'AbortError') {
        throw new ApiException('The request timed out. Please check your connection and try again.');
      }

      throw new ApiException('Unable to reach the server. Please try again.');
    }
  }
}

export const apiClient = new ApiClient();
